### PR TITLE
[Enhancement 2.0] mmdeploy for mmyolo

### PR DESCRIPTION
**Pull request by hanrui1sensetime**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Only support the necessary files. Others are put in mmyolo codebase.

## Modification

- [x] support for external codebase
- [x] import necessary codebase

## BC-breaking (Optional)

None.

## Use cases (Optional)

None.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdeploy/utils/config_utils.py [hunk 1]
```diff
@@ -62,6 +62,18 @@ def get_task_type(deploy_cfg: Union[str, mmengine.Config]) -> Task:
     return Task.get(task)
 
 
+def export_codebase(codebase: str):
+    from aenum import extend_enum
+    try:
+        Codebase.get(codebase)
+    except Exception:
```
Comment (lvhan028) [inline mmdeploy/utils/config_utils.py:69]: I think it's better to print the exception message as well.

File: mmdeploy/utils/config_utils.py [hunk 2]
```diff
@@ -62,6 +62,18 @@ def get_task_type(deploy_cfg: Union[str, mmengine.Config]) -> Task:
     return Task.get(task)
 
 
+def export_codebase(codebase: str):
+    from aenum import extend_enum
```
Comment (lvhan028) [inline mmdeploy/utils/config_utils.py:66]: May add docstring

File: mmdeploy/utils/config_utils.py [hunk 3]
```diff
@@ -62,6 +62,18 @@ def get_task_type(deploy_cfg: Union[str, mmengine.Config]) -> Task:
     return Task.get(task)
 
 
+def export_codebase(codebase: str):
```
Comment (lvhan028) [inline mmdeploy/utils/config_utils.py:65]: I think this function aims at adding a new codebase to `Codebase`. 
So, I think `register_codebase` is more appropriate.

File: mmdeploy/codebase/__init__.py [hunk 4]
```diff
@@ -10,7 +12,7 @@
 }
 
 
-def import_codebase(codebase: Codebase):
+def import_codebase(codebase: Codebase, deploy_cfg: Config = None):
```
Comment (lvhan028) [inline mmdeploy/codebase/__init__.py:15]: May update docstring

File: mmdeploy/utils/config_utils.py [hunk 5]
```diff
@@ -62,6 +62,27 @@ def get_task_type(deploy_cfg: Union[str, mmengine.Config]) -> Task:
...
+def register_codebase(codebase: str) -> Codebase:
+    """Register a new codebase which is not included in Codebase.
+
+    Args:
+        codebase (str): The codebase name.
+
+    Returns:
+        Codebase : An enumeration denotes the codebase type.
+    """
+    from aenum import extend_enum
```
Comment (RunningLeon) [inline mmdeploy/utils/config_utils.py:74]: may add `aenum` into requirements/runtime.txt'

File: mmdeploy/codebase/__init__.py [hunk 6]
```diff
@@ -20,22 +22,35 @@ def import_codebase(codebase: Codebase):
...
     codebase_name = codebase.value
     dependent_library = [codebase_name] + \
         extra_dependent_library.get(codebase, [])
-
+    external_module_list = []
+    if deploy_cfg:
+        external_module_list += deploy_cfg['codebase_config'].get('module', [])
+        dependent_library += deploy_cfg['codebase_config']. \
+            get('extra_dependent_library', [])
+    print(f'debugging what is deploy_cfg: {deploy_cfg}')
```
Comment (RunningLeon) [inline mmdeploy/codebase/__init__.py:36]: remove debug line

File: mmdeploy/codebase/mmocr/deploy/mmocr.py [hunk 7]
```diff
@@ -18,6 +18,15 @@ class MMOCR(MMCodebase):
 
     task_registry = MMOCR_TASK
 
+    @classmethod
+    def import_module(cls):
+        from mmdet.utils.setup_env import \
+            register_all_modules as register_all_modules_mmdet
+        from mmocr.utils.setup_env import \
+            register_all_modules as register_all_modules_mmocr
+        register_all_modules_mmdet(False)
```
Comment (grimoire) [inline mmdeploy/codebase/mmocr/deploy/mmocr.py:27]: Why false?

File: mmdeploy/codebase/__init__.py [hunk 8]
```diff
@@ -1,16 +1,21 @@
...
 
 from mmdeploy.utils import Codebase
 from .base import BaseTask, MMCodebase, get_codebase_class
 
-extra_dependent_library = {
-    Codebase.MMOCR: ['mmdet'],
-    Codebase.MMROTATE: ['mmdet']
-}
 
+def import_codebase_rewriter(codebase_type: Codebase,
```
Comment (grimoire) [inline mmdeploy/codebase/__init__.py:8]: Why do we add another function?
`import_codebase_rewriter` and `import_codebase` always works together.

File: mmdeploy/apis/utils/utils.py [hunk 9]
```diff
@@ -18,7 +18,8 @@ def build_task_processor(model_cfg: mmengine.Config,
         BaseTask: A task processor.
     """
     codebase_type = get_codebase(deploy_cfg)
-    import_codebase(codebase_type)
+    custom_module_list = deploy_cfg['codebase_config'].get('module', [])
```
Comment (RunningLeon) [inline mmdeploy/apis/utils/utils.py:21]: better to use `get_codebase_config` function

File: mmdeploy/utils/config_utils.py [hunk 10]
```diff
@@ -423,3 +423,8 @@ def get_precision(deploy_cfg: Union[str, mmengine.Config]) -> str:
     if backend == Backend.NCNN and 'precision' in deploy_cfg['backend_config']:
         precision = deploy_cfg['backend_config']['precision']
     return precision
+
+
+def get_codebase_external_module(
```
Comment (RunningLeon) [inline mmdeploy/utils/config_utils.py:428]: may add ut for this func @hanrui1sensetime

File: mmdeploy/codebase/mmocr/deploy/mmocr.py [hunk 11]
```diff
@@ -18,6 +18,15 @@ class MMOCR(MMCodebase):
 
     task_registry = MMOCR_TASK
 
+    @classmethod
+    def import_module(cls):
```
Comment (grimoire) [inline mmdeploy/codebase/mmocr/deploy/mmocr.py:22]: As this function is used to invoke `register_all_modules`, I think we can rename it to `register_all_modules` so user won't get confused.

**Review by grimoire — APPROVED** [review]

**Review by RunningLeon — APPROVED** [review]
LGTM

**Pull request merged**